test/extensions/filters/http/ext_authz/config_test.cc

```diff
@@ -3,111 +3,212 @@
+    main_dispatcher_->post([&]() {
+      // Wait until all the workers start to execute the callback.
+      start_counter.Wait();
+      // Notify all the worker to continue.
+      workers_should_fire.Notify();
+    });
+
+    for (Event::DispatcherPtr& dispatcher : worker_dispatchers_) {
+      dispatcher->post([&]() {
+        start_counter.DecrementCount();
+        // Wait for signal from the main thread to start.
+        workers_should_fire.WaitForNotification();
+        work();
+        end_counter.DecrementCount();
+      });
+    }
+    // Wait for all workers to finish the job.
+    end_counter.Wait();
   }
 
-  ExtAuthzFilterConfig factory;
-  ProtobufTypes::MessagePtr proto_config = factory.createEmptyConfigProto();
-  TestUtility::loadFromYaml(
-      fmt::format(grpc_service_yaml, TestUtility::getVersionStringFromApiVersion(api_version)),
-      *proto_config);
-
-  testing::StrictMock<Server::Configuration::MockFactoryContext> context;
-  testing::StrictMock<Server::Configuration::MockServerFactoryContext> server_context;
-  EXPECT_CALL(context, getServerFactoryContext())
-      .WillRepeatedly(testing::ReturnRef(server_context));
-  EXPECT_CALL(context, messageValidationVisitor());
-  EXPECT_CALL(context, clusterManager()).Times(2);
-  EXPECT_CALL(context, runtime());
-  EXPECT_CALL(context, scope()).Times(3);
-
-  Http::FilterFactoryCb cb = factory.createFilterFactoryFromProto(*proto_config, "stats", context);
-  Http::MockFilterChainFactoryCallbacks filter_callback;
-  EXPECT_CALL(filter_callback, addStreamFilter(_));
-  // Expect the raw async client to be created inside the callback.
-  // The creation of the filter callback is in main thread while the execution of callback is in
-  // worker thread. Because of the thread local cache of async client, it must be created in worker
-  // thread inside the callback.
-  EXPECT_CALL(context.cluster_manager_.async_client_manager_, getOrCreateRawAsyncClient(_, _, _, _))
-      .WillOnce(Invoke(
-          [](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool, Grpc::CacheOption) {
-            return std::make_unique<NiceMock<Grpc::MockAsyncClient>>();
-          }));
-  cb(filter_callback);
-
-  Thread::ThreadPtr thread = Thread::threadFactoryForTest().createThread([&context, cb]() {
-    Http::MockFilterChainFactoryCallbacks filter_callback;
-    EXPECT_CALL(filter_callback, addStreamFilter(_));
-    // Execute the filter factory callback in another thread.
-    EXPECT_CALL(context.cluster_manager_.async_client_manager_,
-                getOrCreateRawAsyncClient(_, _, _, _))
-        .WillOnce(Invoke(
-            [](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool,
-               Grpc::CacheOption) { return std::make_unique<NiceMock<Grpc::MockAsyncClient>>(); }));
-    cb(filter_callback);
-  });
-  thread->join();
-}
+protected:
+  void startThreading() {
+    spawnMainThread();
+    spawnWorkerThreads();
```

Comment:
I think you might be able to combine `startThreading()` with` postWorkToAllWorkers()`. In both call-sites you are initializing a factory between those two, but it doesn't look like there'd be anything stopping you from initializing the factory before the call to `startThreading()`.

Reply:
Makes sense to create the factory before threading. but the test class will be more flexible if we don't combine them. For example, we can post the same job several times.

Reply:
Cool cool no doubt no doubt :)

But you weren't actually doing that :)

OK leave them separate then.